utils/src/maybe_encrypted.rs

```diff
@@ -0,0 +1,177 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::marker::PhantomData;
+
+use crypto::{random::make_true_rng, symkey::SymmetricKey};
+use serialization::{Decode, DecodeAll, Encode};
+use zeroize::Zeroizing;
+
+/// A generic type that stores a value as a vector of bytes that are optionally encrypted
+#[derive(Encode, Decode)]
+pub struct MaybeEncrypted<T: Encode + Decode + DecodeAll> {
+    value: Vec<u8>,
+    _phantom: PhantomData<fn() -> T>,
+}
+
+impl<T: Decode + Encode + DecodeAll> MaybeEncrypted<T> {
+    /// Creates a new `MaybeEncrypted` instance with the given value and encryption key.
+    ///
+    /// If the encryption key is `Some`, the value will be encrypted using the key.
+    /// If the encryption key is `None`, the value will be stored as plain bytes.
+    pub fn new(value: &T, encryption_key: &Option<SymmetricKey>) -> Self {
+        match encryption_key {
+            Some(key) => Self::new_encrypted(value, key),
+            None => Self::new_plain(value),
+        }
+    }
+
+    fn new_plain(value: &T) -> Self {
+        Self {
+            value: value.encode(),
+            _phantom: Default::default(),
+        }
+    }
+
+    fn new_encrypted(value: &T, encryption_key: &SymmetricKey) -> Self {
+        Self {
+            value: encryption_key
+                .encrypt(
+                    Zeroizing::new(value.encode()).as_slice(),
+                    &mut make_true_rng(),
+                    None,
+                )
+                .expect("should not fail"),
+            _phantom: Default::default(),
+        }
+    }
+
+    /// Attempts to take the value from the `MaybeEncrypted` instance.
+    ///
+    /// If the encryption key is `Some`, the value will be decrypted using the key.
+    /// If the encryption key is `None`, the value is assumed to be plain bytes and decoded as is.
+    pub fn try_take(
+        self,
+        encryption_key: &Option<SymmetricKey>,
+    ) -> Result<T, crypto::symkey::Error> {
+        match encryption_key {
+            Some(key) => self.try_take_decrypt(key),
+            None => self.try_take_plain(),
+        }
+    }
+
+    fn try_take_plain(self) -> Result<T, crypto::symkey::Error> {
+        T::decode_all(&mut self.value.as_slice()).map_err(|_| {
+            crypto::symkey::Error::DecryptionError(
+                "Could not decode plain content probably it is encrypted".into(),
+            )
+        })
+    }
+
+    pub fn try_take_decrypt(
+        self,
+        encryption_key: &SymmetricKey,
+    ) -> Result<T, crypto::symkey::Error> {
+        encryption_key.decrypt(self.value.as_slice(), None).map(|decrypted_bytes| {
+            T::decode_all(&mut Zeroizing::new(decrypted_bytes).as_slice())
+                .expect("should have been correctly encoded")
```

Comment:
You cannot enforce this invariant based on this design. Given that Decode is derived for MaybeEncrypted, is by implementing Decode yourself and ensuring that any decoded object holds the required invariant.

I recommend that you support decoding errors. Either way, with or without that, it's possible to have invalid objects in MaybeEncrypted.

Reply:
Added new MaybeEncryptedError with decoder or decryption errors.
In Unlock both errors are treated as invalid password, although decoding should not fail unless the DB was corrupted or modified.